base/reduce.jl

```diff
@@ -48,70 +57,55 @@ mul_prod(x::SmallUnsigned,y::SmallUnsigned) = UInt(x) * UInt(y)
     return v
 end
 
-"""
-    mapfoldl(f, op, v0, itr)
-
-Like [`mapreduce`](@ref), but with guaranteed left associativity, as in [`foldl`](@ref).
-`v0` will be used exactly once.
-"""
-mapfoldl(f, op, v0, itr) = mapfoldl_impl(f, op, v0, itr)
-
-"""
-    mapfoldl(f, op, itr)
-
-Like `mapfoldl(f, op, v0, itr)`, but using the first element of `itr` to generate `v0`.
-Specifically, `mapfoldl(f, op, itr)` produces the same result as
-`mapfoldl(f, op, f(first(itr)), drop(itr, 1))`.
-In general, this cannot be used with empty collections (see [`reduce(op, itr)`](@ref)).
-"""
-function mapfoldl(f, op, itr)
+function mapfoldl_impl(f, op, nt::NamedTuple{()}, itr)
     y = iterate(itr)
     if y === nothing
         return Base.mapreduce_empty_iter(f, op, itr, IteratorEltype(itr))
     end
     (x, i) = y
-    v0 = mapreduce_first(f, op, x)
-    mapfoldl_impl(f, op, v0, itr, i)
+    init = mapreduce_first(f, op, x)
+    mapfoldl_impl(f, op, (init=init,), itr, i)
 end
 
-"""
-    foldl(op, v0, itr)
 
-Like [`reduce`](@ref), but with guaranteed left associativity. `v0` will be used
-exactly once.
+"""
+    mapfoldl(f, op, itr; [init])
```

Comment:
I don't think we use brackets to signal that keyword arguments are optional, see e.g. `sum`. I'm not sure whether we should or not, but better be consistent about it.

Reply:
I was following the examples in` titlecase`, `replace`, `replace!` (and I think some others). Let's discuss and deal with whatever is the best documentation pattern here separately.